src/wallet/rpcwallet.cpp

```diff
@@ -305,6 +305,289 @@ UniValue sendtoaddress(const UniValue& params, bool fHelp)
+            // that it's a change address. Ideally we'd have a better way
+            // of checking this by exploring the transaction graph;
+            t_change_dests.insert(item.first);
+        } else {
+            // already accounted for in the address book
+        }
+    }
+
+    /// sprout addresses
+    std::set<libzcash::SproutPaymentAddress> sproutAddresses;
+    pwalletMain->GetSproutPaymentAddresses(sproutAddresses);
+
+    /// sapling addresses
+    std::set<libzcash::SaplingPaymentAddress> saplingAddresses;
+    pwalletMain->GetSaplingPaymentAddresses(saplingAddresses);
+
+    // legacy_random source
+    {
+        // Add legacy randomly generated address records to the result.
+        // This includes transparent addresses generated by the wallet via
+        // the keypool and Sprout addresses for which we have the
+        // spending key.
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "legacy_random");
+        bool hasData = false;
+
+        UniValue random_t(UniValue::VOBJ);
+
+        if (!t_generated_dests.empty()) {
+            UniValue random_t_addrs(UniValue::VARR);
+            for (const CTxDestination& dest : t_generated_dests) {
+                random_t_addrs.push_back(keyIO.EncodeDestination(dest));
+            }
+            random_t.pushKV("addresses", random_t_addrs);
+            hasData = true;
+        }
+
+        if (!t_change_dests.empty()) {
+            UniValue random_t_change_addrs(UniValue::VARR);
+            for (const CTxDestination& dest : t_change_dests) {
+                random_t_change_addrs.push_back(keyIO.EncodeDestination(dest));
+            }
+            random_t.pushKV("change_addresses", random_t_change_addrs);
+            hasData = true;
+        }
+
+        if (!t_generated_dests.empty() || !t_change_dests.empty()) {
+            entry.pushKV("transparent", random_t);
+        }
+
+        if (!sproutAddresses.empty()) {
+            UniValue random_sprout_addrs(UniValue::VARR);
+            for (const SproutPaymentAddress& addr : sproutAddresses) {
+                if (HaveSpendingKeyForPaymentAddress(pwalletMain)(addr)) {
+                    random_sprout_addrs.push_back(keyIO.EncodePaymentAddress(addr));
+                }
+            }
+
+            UniValue random_sprout(UniValue::VOBJ);
+            random_sprout.pushKV("addresses", random_sprout_addrs);
+
+            entry.pushKV("sprout", random_sprout);
+            hasData = true;
+        }
+
+        if (hasData) {
+            ret.push_back(entry);
+        }
+    }
+
+    /// imported source
+    {
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "imported");
+
+        bool hasData = false;
+        {
+            UniValue imported_sapling_addrs(UniValue::VARR);
+            for (const SaplingPaymentAddress& addr : saplingAddresses) {
+                if (GetSourceForPaymentAddress(pwalletMain)(addr) == Imported) {
```

Comment:
Will this catch addresses from imported viewing keys as well?

Reply:
Those lines let you get the ExtFVK corresponding to the address, but do not check whether you _could_ have also obtained the ExtSK. So yes, I think there needs to be an additional check here, otherwise we'll see (at a minimum) the default address for an imported Sapling viewing key in both outputs.